base/multidimensional.jl

```diff
@@ -178,100 +172,164 @@ end  # IteratorsMD
     N = length(I)
     quote
-        @nexprs $N d->(size(A, d) == length(I[d]) || throw(DimensionMismatch("index $d has length $(length(I[d])), but size(A, $d) = $(size(A,d))")))
-        nothing
+        $(Expr(:meta, :inline))
+        @nexprs $N d->(I_d = to_index(I[d]))
+        dest = similar(A, @ncall $N index_shape A I)
+        @ncall $N checksize dest I
+        @ncall $N _unsafe_getindex! linearindexing(dest) dest l A I
+    end
+end
+
+# logical indexing optimization - don't use find (within to_index)
+# Todo: use magic to speed up LinearSlow src
+@inline function _unsafe_getindex(::LinearIndexing, src::AbstractArray, I::AbstractArray{Bool})
+    # Both index_shape and checksize compute sum(I); manually hoist it out
+    N = sum(I)
+    dest = similar(src, (N,))
+    size(dest) == (N,) || throw(DimensionMismatch())
+    c = 1
+    for i = 1:length(I)
+        if unsafe_getindex(I, i)
+            unsafe_setindex!(dest, src[i], c)
+            c += 1
+        end
     end
+    dest
 end
 
-@inline unsafe_getindex(v::BitArray, ind::Int) = Base.unsafe_bitgetindex(v.chunks, ind)
+# Each time we call out to the user getindex, we add another function call to
+# the stack. So within the _functions, we only call back out once we've reduced
+# the call the whole way to the canonical override. Performance here is very
+# sensitive.
+_getindex(::LinearFast, A::AbstractArray, index::CartesianIndex{1}) = getindex(A, index[1])
+stagedfunction _getindex{N}(l::LinearFast, A::AbstractArray, index::CartesianIndex{N})
+    :($(Expr(:meta, :inline)); @ncall $N _getindex l A d->index[d])
+end
+stagedfunction _getindex{N}(::LinearSlow, A::AbstractArray, index::CartesianIndex{N})
+    :($(Expr(:meta, :inline)); @ncall $N getindex A d->index[d])
+end
+_unsafe_getindex(::LinearFast, A::AbstractArray, index::CartesianIndex{1}) = unsafe_getindex(A, index[1])
+stagedfunction _unsafe_getindex{N}(l::LinearFast, A::AbstractArray, index::CartesianIndex{N})
+    :($(Expr(:meta, :inline)); @ncall $N _unsafe_getindex l A d->index[d])
+end
+stagedfunction _unsafe_getindex{N}(::LinearSlow, A::AbstractArray, index::CartesianIndex{N})
+    :($(Expr(:meta, :inline)); @ncall $N unsafe_getindex A d->index[d])
+end
 
-@inline unsafe_setindex!{T}(v::Array{T}, x::T, ind::Int) = (@inbounds v[ind] = x; v)
-@inline unsafe_setindex!{T}(v::AbstractArray{T}, x::T, ind::Int) = (v[ind] = x; v)
-@inline unsafe_setindex!(v::BitArray, x::Bool, ind::Int) = (Base.unsafe_bitsetindex!(v.chunks, x, ind); v)
-@inline unsafe_setindex!{T}(v::AbstractArray{T}, x::T, ind::Real) = unsafe_setindex!(v, x, to_index(ind))
+# Indexing with just one array is always linear in the source, fast or slow
+# TODO: Use magic to speed up repeated ind2sub calls
+@inline function _unsafe_getindex!(::LinearFast, dest::AbstractArray, ::LinearIndexing, src::AbstractArray, I::AbstractArray)
+    for i=1:length(dest)
```

Comment:
How about a construct like this:
```julia
idest = 0
for isrc in I
    idest += 1
    unsafe_setindex!(dest, getindex(src, isrc), idest)
end
```
That way you don't implicitly assume that `I` has fast linear indexing.